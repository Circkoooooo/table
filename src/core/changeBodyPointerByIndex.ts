import changeBodyCursor from "./tools/changeBodyCursor"
import { IndexType } from "./types/table.types"

/**
 * type参数为reset则改变cursor为pointer；如果是calc，则提取第二个参数index，判断是否是pointer
 * @param type reset | calc
 */
const changeBodyPointerByIndex = <T extends "reset" | "calc">(type: T, index?: T extends "calc" ? IndexType : never) => {
	if (type === "reset") {
		changeBodyCursor("auto")
	} else if (type === "calc" && index !== undefined) {
		const { rowIndex, columnIndex } = index
		if (rowIndex === 0 || columnIndex === 0) {
			changeBodyCursor("pointer")
		} else {
			changeBodyCursor("auto")
		}
	}
}

export default changeBodyPointerByIndex
