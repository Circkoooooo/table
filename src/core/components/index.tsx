import { TableFrame } from "../styled/Table-styled"
import { TableMain } from "./TableMain"
import { useEffect } from "react"

const Table = () => {
	useEffect(() => {
		const contextMenuListenerCallback = (event: MouseEvent) => {
			event.preventDefault()
		}

		window.addEventListener("contextmenu", contextMenuListenerCallback)

		return () => {
			window.removeEventListener("contextmenu", contextMenuListenerCallback)
		}
	}, [])

	return (
		<>
			<TableFrame>
				<TableMain />
			</TableFrame>
		</>
	)
}

export default Table
