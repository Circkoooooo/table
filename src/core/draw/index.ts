export type Coordinate = {
	x: number
	y: number
}

export type DrawLineProperty = {
	lineColor?: string
	lineWidth?: number
	maxRenderRowCount?: number
	maxRenderColumnCount?: number
}
/**
 *
 * @param canvasId id of canvas target
 * @param initialWidth canvas width the first render
 * @param initialHeight canvas height the first render
 * @returns
 */
const CustomCanvas = (_canvasTarget: HTMLCanvasElement) => {
	const canvasState = {
		canvasTarget: _canvasTarget,
		_canvasContext: _canvasTarget.getContext("2d"),
		canvasLineWidth: 1,
		canvasStrokeColor: "#000000",
		devicePixelRatio: 1,
		pixelOffset: 0.5,
		fontSize: 16,
	}

	const getCanvasContext = () => {
		return canvasState._canvasContext
	}

	const updateStrokeColor = (lineColor?: DrawLineProperty["lineColor"]) => {
		if (!lineColor || lineColor === canvasState.canvasStrokeColor) return
		const ctx = getCanvasContext()
		ctx && (ctx.strokeStyle = lineColor)

		canvasState.canvasStrokeColor = lineColor
	}

	const updateCanvasLineWidth = (lineWidth?: DrawLineProperty["lineWidth"]) => {
		if (!lineWidth) return
		const ctx = getCanvasContext()

		canvasState.canvasLineWidth = lineWidth
		ctx && (ctx.lineWidth = canvasState.canvasLineWidth)

		return canvasState.canvasLineWidth
	}

	const recordDpr = () => {
		canvasState.devicePixelRatio = window.devicePixelRatio
		return canvasState.devicePixelRatio
	}

	const getDpr = () => {
		return canvasState.devicePixelRatio
	}

	const getPixelOffset = (lineWidth: number) => {
		const ctx = getCanvasContext()
		if (!ctx) return null
		return lineWidth % 2 === 0 ? 0 : 0.5
	}

	const measureText = (text: string) => {
		const ctx = getCanvasContext()
		if (!ctx) return null
		return ctx.measureText(text)
	}

	const getTextHeight = (fontSize: number = canvasState.fontSize) => {
		const ctx = getCanvasContext()
		if (!ctx) return null

		let size = fontSize
		if (ctx.textBaseline === "hanging" || ctx.textBaseline === "top") {
			size = fontSize
		} else if (ctx.textBaseline === "middle") {
			size = fontSize / 2
		}

		return size
	}

	/**
	 * 调用来实现线条的绘制。
	 *
	 * @returns
	 * beginPath: 线条绘制前调用
	 *
	 * markLine: 在指定位置标记线条来为渲染做准备
	 *
	 * strokeLine：渲染标记好的线条
	 */
	const drawLine = () => {
		const context = getCanvasContext()

		const beginPath = () => {
			if (!context) return
			context.beginPath()
		}

		const markLine = (start: Coordinate, end: Coordinate) => {
			if (!context) return

			const offset = getPixelOffset(canvasState.canvasLineWidth) ?? 0
			context.moveTo(start.x + offset, start.y + offset)
			context.lineTo(end.x + offset, end.y + offset)
		}

		const strokeLine = () => {
			if (!context) return
			context.stroke()
		}

		const closePath = () => {
			if (!context) return
			context.closePath()
		}

		return {
			beginPath,
			markLine,
			strokeLine,
			closePath,
		}
	}

	/**
	 * 调用来绘制一个文本
	 * 默认对齐为横向：start 纵向top
	 * 默认字体为16
	 * @returns
	 */
	const drawText = () => {
		const context = getCanvasContext()

		const fillText = (
			text: string,
			x: number,
			y: number,
			fontSize: number = canvasState.fontSize,
			textAlign: CanvasTextAlign = "start",
			textBaseline: CanvasTextBaseline = "top",
			renderWidth?: number
		) => {
			if (!context) return

			const offset = getPixelOffset(canvasState.canvasLineWidth) ?? 0
			context.font = `${fontSize}px Microsoft Yahei`
			context.textAlign = textAlign
			context.textBaseline = textBaseline

			const measureFillText = measureText(text)
			//换行
			const splitTextArr = []
			if (renderWidth && text !== "") {
				// renderWidth来作为最大的渲染宽度，实现换行
				let currentIndex = 0
				let currentSplitStartIndex = 0

				while (currentIndex < text.length) {
					const currentText = text.slice(currentSplitStartIndex, currentIndex + 1)

					const measureCurrentText = measureText(currentText)
					const currentMeasureWidth = measureCurrentText?.width || 0

					if (currentMeasureWidth > renderWidth) {
						const splitText = text.slice(currentSplitStartIndex, currentIndex)
						splitTextArr.push(splitText)
						currentSplitStartIndex = currentIndex
					}

					if (currentIndex === text.length - 1) {
						splitTextArr.push(text.slice(currentSplitStartIndex))
					}

					currentIndex++
				}
			} else {
				splitTextArr.push(text)
			}

			//渲染每一行
			splitTextArr.forEach((text, index) => {
				/**
				 * (measureFillText?.actualBoundingBoxAscent || 0) + (measureFillText?.actualBoundingBoxDescent || 0) 计算出文字高度
				 * 单纯用文字高度来做偏移，文字显示可能会重叠，所以增加2个像素的额外偏移
				 */
				const extraOffsetToPreLine = 2
				const textOffset = index * (extraOffsetToPreLine + (measureFillText?.actualBoundingBoxAscent || 0) + (measureFillText?.actualBoundingBoxDescent || 0))
				context.fillText(text, x + offset, y + offset + textOffset)
			})
		}

		return { fillText }
	}

	/**
	 * 绘制矩形
	 */
	const fillRect = (x: number, y: number, endX: number, endY: number, color: string = "#000000") => {
		const context = getCanvasContext()
		if (!context) return

		const oldFillStyle = context.fillStyle
		context.fillStyle = color
		context.fillRect(x, y, endX, endY)
		context.fillStyle = oldFillStyle
	}

	const clipRect = (startX: number, startY: number, width: number, height: number) => {
		const context = getCanvasContext()
		if (!context) return
		context.save()
		context.beginPath()
		context.rect(startX, startY, width, height)
		context.clip()
	}

	const restoreClip = () => {
		const context = getCanvasContext()
		context?.restore()
	}

	/**
	 * 接受宽度和高度，刷新当前绑定canvas的宽度和高度，【适配dpr缩放】
	 * @param initialWidth
	 * @param initialHeight
	 * @returns
	 */
	const updateSize = (initialWidth: number, initialHeight: number) => {
		const canvas = canvasState.canvasTarget
		canvas.width = initialWidth
		canvas.height = initialHeight

		const ctx = getCanvasContext()
		if (!canvas || !ctx) return

		const dpr = recordDpr()

		canvas.style.width = `${initialWidth / dpr}px`
		canvas.style.height = `${initialHeight / dpr}px`
	}

	return {
		drawLine,
		drawText,
		updateSize,
		getDpr,
		getPixelOffset,
		measureText,
		getTextHeight,
		updateStrokeColor,
		updateCanvasLineWidth,
		clipRect,
		restoreClip,
		fillRect,
	}
}

export { CustomCanvas }
export type CustomCanvasType = ReturnType<typeof CustomCanvas>
