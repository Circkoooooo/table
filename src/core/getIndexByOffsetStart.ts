import { TableRowColumnCellConfigExtraSizeConfig } from "./redux/canvas/canvasSlice.types"

/**
 * 第一个单元格包含边框，所以全逻辑宽度以内都为该单元格的索引
 * 后面所有单元格之间由于重叠了边框，所以需要以逻辑宽度减去边框宽度
 * @param offsetStartNum
 * @param lineWidth
 * @param logicSize
 * @param rowColumnCellExtraSizeConfig
 * @returns
 */
export const getIndexByOffsetStart = (offsetStartNum: number, lineWidth: number, logicSize: number, rowColumnCellExtraSizeConfig: TableRowColumnCellConfigExtraSizeConfig) => {
	let ofsStartTemp = offsetStartNum
	let index = 0

	while (ofsStartTemp >= logicSize) {
		const currentIndex = index
		ofsStartTemp -= logicSize - lineWidth + (rowColumnCellExtraSizeConfig.find((item) => item.index === currentIndex)?.value || 0)
		index++
	}

	return index
}
