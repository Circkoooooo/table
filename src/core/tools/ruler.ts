import deepClone from "../../tools/deepClone"
import { GetColumnLabel, PickValidTypeAsciiCode, ResolvePrefixArray } from "./types"

/**
 * 对数组中的值从后往前进行自增进位
 * 如果是空数组[]，则传入最小的值minPrefixCode，否则对最后的值进行自增，最后一个值如果大于maxPrefixCode则将它置为minPrefixCode，向前进位。
 *
 * 例如当minPrefixCode=65，maxPrefixCode=90。用=>来表示这个函数的调用
 * 那么当prefixArray为如下值的时候，经过处理后prefixArray中的值变化如下
 * prefixArray=[65]  =>  [66]
 * prefixArray=[90]     =>  [65,65]
 * prefixArray=[65,65]  =>  [65,66]
 * prefixArray=[]       =>  [65]
 * prefixArray=[90,90]  =>  [65,65,65]
 *
 * @param prefixArray
 * @param minPrefixCode
 * @param maxPrefixCode
 */
export const resolvePrefixArray: ResolvePrefixArray = (prefixArray: number[], minPrefixCode: number, maxPrefixCode: number) => {
	const resultPrefixArray: number[] = deepClone(prefixArray)

	if (prefixArray.length === 0) {
		resultPrefixArray[0] = minPrefixCode
	} else {
		if (resultPrefixArray.length === 1) {
			if (resultPrefixArray[0] === maxPrefixCode) {
				resultPrefixArray[0] = minPrefixCode
				resultPrefixArray.unshift(minPrefixCode)
			} else {
				resultPrefixArray[0]++
			}
		} else {
			for (let i = resultPrefixArray.length - 1, carry = false; i >= 0; i--) {
				//最后一个索引
				if (i === resultPrefixArray.length - 1) {
					if (resultPrefixArray[i] === maxPrefixCode) {
						carry = true
						resultPrefixArray[i] = minPrefixCode
					} else {
						resultPrefixArray[i]++
					}
				} else {
					if (carry) {
						carry = false

						let isLastPrefixCarry = false
						if (resultPrefixArray[i] === maxPrefixCode) {
							isLastPrefixCarry = true
							resultPrefixArray[i] = minPrefixCode
						} else {
							resultPrefixArray[i]++
						}

						if (isLastPrefixCarry && i === 0) {
							resultPrefixArray.unshift(minPrefixCode)
						}
					}
				}
			}
		}
	}
	return resultPrefixArray
}

const pickValidTypeAsciiCode: PickValidTypeAsciiCode = (asciiCode: number | string) => {
	if (typeof asciiCode === "number") {
		return asciiCode
	}

	return asciiCode.charCodeAt(0)
}

/**
 * 生成ascii码区间内符合A - Z .. AA - AZ .. AAA AAB - ZZZ ...规则的取值数组
 *
 */
export const getColumnLabel: GetColumnLabel = (columnCount: number, asciiMin: number | string = 65, asciiMax: number | string = 90) => {
	const asciiLimit: number[] = []
	const result: string[] = []

	asciiLimit.push(pickValidTypeAsciiCode(asciiMin))
	asciiLimit.push(pickValidTypeAsciiCode(asciiMax))

	//前缀
	const previousPrefixCode = {
		prefixes: [] as number[],
		needAddPrefix: false,
	}

	// 遍历columnCount次，push的内容为 A B .... Z AA AB AC .... BA ..ZZ ...AAA AAB ...
	for (let i = asciiLimit[0]; i < asciiLimit[0] + columnCount; i++) {
		const currentLoopValue = ((i - asciiLimit[0]) % (asciiLimit[1] - asciiLimit[0] + 1)) + asciiLimit[0]

		//判断是否存在要在赋值`needAddPrefix`变量之前，因为这是在下一轮循环中判断的。
		//如果有前缀标记，处理前缀数组
		if (previousPrefixCode.needAddPrefix) {
			previousPrefixCode.needAddPrefix = false
			previousPrefixCode.prefixes.splice(0)
			previousPrefixCode.prefixes.push(...resolvePrefixArray(previousPrefixCode.prefixes, asciiLimit[0], asciiLimit[1]))
		}

		//判断是否为最后一个值，如果是，记录前缀标记
		if (currentLoopValue === asciiLimit[1]) {
			previousPrefixCode.needAddPrefix = true
		}

		//组合前缀和当前索引
		result.push(previousPrefixCode.prefixes.map((item) => String.fromCharCode(item)).join("") + String.fromCharCode(currentLoopValue))
	}
	return result
}

export const getRowLabel = (count: number, offset = 0): string[] => {
	const result: string[] = []

	result.push(...Array.from({ length: count }, (val, key) => key + 1 + offset).map((item) => item.toString()))

	return result
}
