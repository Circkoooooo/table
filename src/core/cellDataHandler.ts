import deepClone from "../tools/deepClone"
import { getColumnLabel, getRowLabel } from "./ruler"

interface FullTableEmptyDataParams {
	rowNum: number
	columnNum: number
}

export type CellDataElement = string | number | null

export type CellData = CellDataElement[][]

export interface WithRulerCellData {
	data: CellData
	info: {
		rowLength: number
		columnLength: number
	}
}
/**
 * 根据row column数目返回一个矩阵 r * c
 * @param params
 */
export const createEmptyCellData = (params: FullTableEmptyDataParams): CellData => {
	const { rowNum, columnNum } = params

	return Array.from(Array(rowNum), () => new Array(columnNum).fill(null))
}

/**
 * 在空二维数组的基础上添加行标和列标（深克隆）
 * @param emptyRulerCellData
 */
export const createRulerCellData = (emptyRulerCellData: CellData): WithRulerCellData => {
	const tempEmptyRulerCellData = deepClone(emptyRulerCellData)

	const result = {
		data: tempEmptyRulerCellData,
		info: {
			rowLength: tempEmptyRulerCellData.length,
			columnLength: (tempEmptyRulerCellData[0] && tempEmptyRulerCellData[0].length) || 0,
		},
	}

	if (tempEmptyRulerCellData.length === 0) return result

	const rowLength = tempEmptyRulerCellData.length
	const columnLength = tempEmptyRulerCellData[0].length

	const columnLabels = getColumnLabel(columnLength, 65, 90)
	const rowLabels = getRowLabel(rowLength, 0)

	const resolvedColumnCellData = [...tempEmptyRulerCellData].map((item, i) => {
		item.unshift(rowLabels[i])
		return item
	})

	//inject one row into
	const rulerCellData = [[null, ...columnLabels], ...resolvedColumnCellData]

	return {
		data: rulerCellData,
		info: {
			rowLength: rulerCellData.length,
			columnLength: rulerCellData[0].length,
		},
	}
}
