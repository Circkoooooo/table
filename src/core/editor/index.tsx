import { TableColumnHeader, TableDataFrame, TableDataRow, TableFrame, TableRowAndDataFrame, TableRowAndDataRowFlex, TableRowHeader } from "./Table-styled"
import React, { useCallback, useEffect, useRef, useState } from "react"
import { getLabel, getRowLabel } from "./Ruler"
import { arrToObject } from "../../tools/arrToObject"
import Cell from "./Cell"
import HintBorder, { HintBorderRef } from "./HintBorder"

interface ColumnRulerProps {
	columnCount?: number
}

interface CurrentSelectCellInfo {
	selectRowIndex: number
	selectColumnIndex: number
	oldSelectSell?: HTMLDivElement
	newTargetTable?: AbstractTableElementType[][]
}

type AbstractTableElementType = string | number | undefined | null

const [EDIT_ATTRIBUTE, EDIT_ATTRIBUTE_VALUE] = ["contentEditable", "true"]

const tables: AbstractTableElementType[][] = [
	[1, 2, 3, 4, 5, 6],
	[3, 4, 5, 6, 7, 8],
	[1, 2, 3, 4, 5, 6],
	[3, 4, 5, 6, 7, 8],
	[4, 5, 6, 7, 8, 8],
	[1, 2, 3, 4, 5, 6],
	[3, 4, 5, 6, 7, 8],
	[4, 5, 6, 7, 8, 8],
	[1, 2, 3, 4, 5, 6],
	[3, 4, 5, 6, 7, 8],
	[4, 5, 6, 7, 8, 8],
	[1, 2, 3, 4, 5, 6],
	[3, 4, 5, 6, 7, 8],
	[4, 5, 6, 7, 8, 8],
	[1, 2, 3, 4, 5, 6],
	[3, 4, 5, 6, 7, 8],
	[4, 5, 6, 7, 8, 8],
	[1, 2, 3, 4, 5, 6],
]

const Table: React.FC<ColumnRulerProps> = () => {
	const [tableAddition, setTableAddition] = useState<{
		rowLabels: string[]
		columnLabels: string[]
	}>({ rowLabels: [], columnLabels: [] })

	const [targetTables, setTargetTables] = useState<AbstractTableElementType[][]>()
	const currentSelectCell = useRef<CurrentSelectCellInfo | null>(null) //click后记录
	const hintBorder = useRef<HintBorderRef>(null)

	/**
	 * 获取当前数据状态下的的label
	 */
	const getCurrentTableAddition = () => {
		return {
			rowLabels: getRowLabel(26 < tables.length ? tables.length : 26, 0),
			columnLabels: tables[0] !== undefined ? getLabel(26 < tables[0].length ? tables[0].length : 26, "A", "Z") : [],
		}
	}

	/**
	 * 根据table的行数和列数生成填充0的表格，然后把table中的内容复制过来
	 */
	const buildMaskTables = useCallback(() => {
		return Array.from({ length: tableAddition.rowLabels.length }, (v, row) => {
			return Array.from({ length: tableAddition.columnLabels.length }, (v, column) => {
				if (row < tables.length && column < tables[row].length) {
					return tables[row][column]
				}
				return undefined
			})
		})
	}, [tableAddition.rowLabels.length, tableAddition.columnLabels.length])

	const { handleCellMouseDown, handleCellBlur, handleCellInput } = (function () {
		return {
			handleCellMouseDown: ({ event, row, column }: { event: React.MouseEvent<HTMLDivElement>; row: number; column: number }) => {
				event.preventDefault()
				if (!targetTables || targetTables.length <= row || targetTables[row].length <= column) {
					throw new Error("Error, it seems that the table has not been rendered.")
				}

				//focus target
				const target = event.currentTarget
				target.setAttribute(EDIT_ATTRIBUTE, EDIT_ATTRIBUTE_VALUE)
				currentSelectCell.current?.oldSelectSell?.blur()
				currentSelectCell.current && (currentSelectCell.current.oldSelectSell = target)

				if (currentSelectCell.current?.selectRowIndex === row && currentSelectCell.current.selectColumnIndex === column) {
					target.focus()
				}

				//record current selected target cell without updating.
				currentSelectCell.current = {
					...currentSelectCell.current,
					selectRowIndex: row,
					selectColumnIndex: column,
				}

				hintBorder.current?.changeIndex(row, column)
			},
			handleCellBlur: (event: React.FocusEvent<HTMLDivElement>) => {
				const target = event.currentTarget

				if (target.hasAttribute(EDIT_ATTRIBUTE)) {
					const target = event.currentTarget

					target.removeAttribute(EDIT_ATTRIBUTE)
				}

				currentSelectCell.current?.newTargetTable && updateTargetTable(currentSelectCell.current.newTargetTable)
			},
			handleCellInput: (target: HTMLElement) => {
				if (!currentSelectCell.current || !targetTables) return

				const { selectRowIndex, selectColumnIndex } = currentSelectCell.current
				const tempTargetTables = [...targetTables] //对所有数组重新改变引用，引起useState的更新

				tempTargetTables[selectRowIndex][selectColumnIndex] = target.innerText
				currentSelectCell.current.newTargetTable = tempTargetTables //记录最新修改的值
			},
		}
	})()

	/**
	 * 对表格内容进行更新
	 * 1. 更新state
	 * 2. 清除ref为null
	 */
	const updateTargetTable = (newTargetTable: AbstractTableElementType[][]) => {
		setTargetTables(newTargetTable)
		currentSelectCell.current = null
		console.log("update", newTargetTable)
	}

	useEffect(() => {
		setTargetTables(buildMaskTables())

		if (tableAddition.rowLabels.length !== tables.length || tableAddition.columnLabels.length !== tables[0].length) {
			setTableAddition(getCurrentTableAddition())
		}
	}, [buildMaskTables, tableAddition.columnLabels.length, tableAddition.rowLabels.length])

	//列表头
	const RenderColumnHeader = useCallback(() => {
		return (
			<>
				<Cell attrs={{ ...arrToObject(["top", "left", "lineRight"]) }}></Cell>
				{tableAddition.columnLabels.map((value, index) => {
					const props = ["left", "rop", "top", "dark", "lineRight"]

					if (index === tableAddition.columnLabels.length - 1) {
						props.push("right")
					}

					return (
						<Cell attrs={{ ...arrToObject(props) }} key={value}>
							{value}
						</Cell>
					)
				})}
			</>
		)
	}, [tableAddition.columnLabels])

	//行表头和行列功能格
	const RenderRowHeader = useCallback(() => {
		return (
			<>
				{tableAddition.rowLabels.map((value, index) => {
					const props = []
					if (index !== tableAddition.rowLabels.length - 1) {
						props.push("left", "dark")
					} else {
						props.push("left", "bottom", "dark")
					}

					props.push("top")

					return (
						<Cell attrs={{ ...arrToObject(props) }} key={value}>
							{value}
						</Cell>
					)
				})}
			</>
		)
	}, [tableAddition.rowLabels])

	const RenderTableData = () => {
		if (!targetTables) return null

		return (
			<>
				{/* {targetTables && */}
				{targetTables.map((item, row) => {
					return (
						<TableDataRow key={tableAddition.rowLabels[row]}>
							{item.map((value, column) => {
								const key = `${tableAddition.columnLabels[column]}${tableAddition.rowLabels[row]}`
								const props = []

								const eventProps = {
									onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => {
										handleCellMouseDown({
											event,
											row,
											column,
										})
									},
									onInput: (event: React.FormEvent<HTMLDivElement>) => handleCellInput(event.currentTarget),
									onBlur: (event: React.FocusEvent<HTMLDivElement>) => handleCellBlur(event),
								}

								if (row !== tableAddition.rowLabels.length - 1) {
									props.push("left")
								} else {
									props.push("left", "bottom")
								}

								if (column === item.length - 1) {
									props.push("right")
								}

								props.push("top")

								return (
									<Cell
										attrs={{
											...arrToObject(props),
											...eventProps,
										}}
										key={key}
									>
										{value}
									</Cell>
								)
							})}
						</TableDataRow>
					)
				})}
			</>
		)
	}

	return (
		<>
			<TableFrame>
				<TableColumnHeader>
					<RenderColumnHeader />
				</TableColumnHeader>

				<TableRowAndDataFrame>
					<TableRowAndDataRowFlex>
						<TableRowHeader>
							<RenderRowHeader />
						</TableRowHeader>
						<TableDataFrame>
							<RenderTableData />
							<HintBorder
								ref={hintBorder}
								{...{
									maxRowIndex: tableAddition.rowLabels.length,
									maxColumnIndex: tableAddition.columnLabels.length,
								}}
							/>
						</TableDataFrame>
					</TableRowAndDataRowFlex>
				</TableRowAndDataFrame>
			</TableFrame>
		</>
	)
}

export default Table
