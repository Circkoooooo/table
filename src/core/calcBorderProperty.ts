import { CellData } from "./cellDataHandler"

export type BorderProperty = {
	top?: boolean
	right?: boolean
	bottom?: boolean
	left?: boolean
}

const calcBorderProperty = (cellData: CellData, rowNum: number, columnNum: number): BorderProperty[][] => {
	return cellData.map((row, rowIndex) => {
		return row.map((_, columnIndex) => {
			const property = {
				top: false,
				right: false,
				bottom: false,
				left: false,
			}

			if (rowIndex === 0) {
				if (columnIndex === 0) {
					property.top = true
					property.right = true
					property.bottom = true
					property.left = true
				} else {
					property.top = true
					property.bottom = true
					property.right = true
				}
			} else {
				// last column and not last row
				if (columnIndex === 0) {
					// property.top = true
					property.right = true
					property.bottom = true
					property.left = true
				} else {
					//neither last row nor last column
					Object.assign(property, {
						right: true,
						bottom: true,
					})
				}
			}

			return property
		})
	})
}

export default calcBorderProperty
