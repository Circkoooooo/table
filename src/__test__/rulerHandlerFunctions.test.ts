import { getColumnLabel, getRowLabel, resolvePrefixArray } from "../core/ruler"

describe("ruler", () => {
	const resultA2Z = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z"]

	it("prefix generator", () => {
		expect(getColumnLabel(26)).toMatchObject(resultA2Z)
		expect(getColumnLabel(26, 65, 90)).toMatchObject(resultA2Z)
		expect(getColumnLabel(27, 65, 90)).toMatchObject([...resultA2Z, "AA"])
		expect(getColumnLabel(26, "A", "Z")).toMatchObject(resultA2Z)
	})

	it("typeof asciiMin is valid type", () => {
		function validType() {
			getColumnLabel(26, null as unknown as any, "Z")
		}
		expect(validType).toThrow(new Error("Type of null is not a valid type"))
	})

	it("typeof asciiMax is valid type", () => {
		function validType() {
			getColumnLabel(26, 90, null as unknown as any)
		}
		expect(validType).toThrow(new Error("Type of null is not a valid type"))
	})

	it("snapshot", () => {
		expect(getColumnLabel(26)).toMatchSnapshot()
	})
})

describe("prefix array", () => {
	it("Empty prefix Array. Returns a array includes a min code.", () => {
		const targetPrefixArray = resolvePrefixArray([], 65, 90)
		expect(targetPrefixArray).toMatchObject([65])
	})

	it("Prefix array [mincode]. Returns [mincode + 1]", () => {
		const targetPrefixArray = resolvePrefixArray([65], 65, 90)
		expect(targetPrefixArray).toMatchObject([66])
	})

	it("Prefix array [maxcode]. Returns [mincode, mincode]", () => {
		const targetPrefixArray = resolvePrefixArray([90], 65, 90)
		expect(targetPrefixArray).toMatchObject([65, 65])
	})

	it("Prefix array [code, code(!maxcode)]. Returns [code, code + 1]", () => {
		const targetPrefixArray = resolvePrefixArray([90, 66], 65, 90)
		expect(targetPrefixArray).toMatchObject([90, 67])
	})

	it("Prefix array [code, maxcode]. the first and the last need carry", () => {
		const targetPrefixArray = resolvePrefixArray([90, 90], 65, 90)
		expect(targetPrefixArray).toMatchObject([65, 65, 65])
	})

	it("Prefix array [code, maxcode-1]. the first not carry", () => {
		const targetPrefixArray = resolvePrefixArray([65, 90], 65, 90)
		expect(targetPrefixArray).toMatchObject([66, 65])
	})
})

describe("getRowLabel", () => {
	const expectedRowLabel = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26"]
	const expectedRowLabelOffset2 = ["3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28"]

	it("row label", () => {
		const rowLable = getRowLabel(26)
		expect(rowLable).toMatchObject(expectedRowLabel)

		expect(rowLable).toMatchSnapshot()
	})

	it("not default offset", () => {
		const rowLable = getRowLabel(26, 2)
		expect(rowLable).toMatchObject(expectedRowLabelOffset2)

		expect(rowLable).toMatchSnapshot()
	})
})
