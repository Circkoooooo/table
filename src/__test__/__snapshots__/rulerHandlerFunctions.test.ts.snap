// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRowLabel not default offset 1`] = `
Array [
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
  "13",
  "14",
  "15",
  "16",
  "17",
  "18",
  "19",
  "20",
  "21",
  "22",
  "23",
  "24",
  "25",
  "26",
  "27",
  "28",
]
`;

exports[`getRowLabel row label 1`] = `
Array [
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
  "10",
  "11",
  "12",
  "13",
  "14",
  "15",
  "16",
  "17",
  "18",
  "19",
  "20",
  "21",
  "22",
  "23",
  "24",
  "25",
  "26",
]
`;

exports[`ruler snapshot 1`] = `
Array [
  "A",
  "B",
  "C",
  "D",
  "E",
  "F",
  "G",
  "H",
  "I",
  "J",
  "K",
  "L",
  "M",
  "N",
  "O",
  "P",
  "Q",
  "R",
  "S",
  "T",
  "U",
  "V",
  "W",
  "X",
  "Y",
  "Z",
]
`;
